man/showattr.8
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.41.2.
.TH SHOWATTR "8" "May 2013" "showattr  -- shows vserver specific file attributes" "System Administration"
.SH NAME
showattr \- shows vserver specific file attributes
.SH SYNOPSIS
.B showattr
[\fI-Radx\fR] [\fI--\fR] \fI<file>*\fR
.SH DESCRIPTION
.IP
Options:
.TP
\fB\-R\fR
\&...  recurse through directories
.TP
\fB\-a\fR
\&...  display files starting with '.' also
.TP
\fB\-d\fR
\&...  list directories like other files instead of listing
.IP
their content
.TP
\fB\-x\fR
\&...  do not cross filesystems
.PP
Please report bugs to vserver@list.linux\-vserver.org
.PP
This program is part of util\-vserver
.SH COPYRIGHT
Copyright \(co 2004 Enrico Scholz
This program is free software; you may redistribute it under the terms of
the GNU General Public License.  This program has absolutely no warranty.


NEWS
version 0.30.215
================

	- added support for using libnss instead of beecrypt for
	vhashify's hash calculation.  libnss has bad SHA1 performance on
	i386, but is a) maintained and b) gives better performance with
	all other hashes and on x86_64.

	- added tunctl, a small program to setup tun/tap interfaces, with
	support in the configuration.

	- vattribute and nattribute now support a --get operation, which
	parses the flags and capabilities into a human-readable form.

	- vspace supports all the spaces currently available.

	- vmemctrl controls the OOM bias of context.

	- dynamic contexts are unbroken.

	- hopefully all rpm-fake races have been fixed.

	- flock has been replaced by lockf, which works on NFS.

	- the clone build method now supports exclude lists.

	- vmount handles mounting into a guest.

	- various fixes for the Gentoo-scripts, thanks to Benedikt Böhm.

	- support for yum's persistdir has been added.

	- per-guest device maps are now supported.

	- vserver ... build now sets the barrier on the parent directory.

	- fsck is now supported for the guest's fstab.

	- symlinks are now handled on all levels in namespace cleanup.

	- the debootstrap build method now supports debootstrap 1.0.8.

	- a documentation update, thanks to Micah Anderson.

	- use numeric uids/gids for the template build method.

	- add support for the alpine and f8 distributions.
	
version 0.30.214
================
	- patches for yum-3.2.

	- various fixes to allow for *SuSE guests.

	- added Fedora 7.

	- replace file with h2ext for build -m template.

	- add exclude list support for vclone.

	- include debootstrap scripts for a lot of distributions.

	- fixes to allow the Gentoo initscripts to be used with baselayout 2.

	- remove disk limit caches on vserver ... delete.

	- added support for the new APIs in 2.2 and 2.3, in particular the
	netv2 API, allowing native IPv6 support.

	- added support for a separate filesystem tag.

version 0.30.213
================
	- added support for the device mapping feature present in development
	kernels.

	- further improved Gentoo support, again thanks to Benedikt Böhm.

	- support persistent contexts.

	- add another initscript that creates directories, configures vshelper
	as well as kills all contexts.

	- new vserver-stat implementation, which uses the accounting APIs
	available in Linux-VServer 2.2.

	- added support for setting sysctl values in the guest configuration.

	- added a clone build method, which simply links already unified files.

	- implement dynamic contexts in userspace, generating a static one
	during vserver ... build.

	- fix the endless loop in vlogin.

	- fix Ubuntu builds.

	- improved support for Debian Etch, thanks to Benedikt Böhm.

	- added CentOS 5.

	- use the state change helper to try to cleanup after us if the guest
	exited on its own.

version 0.30.212
================
	- support for the Linux-VServer 2.1 API has been implemented.

	- the scheduler configuration has been split up into multiple files.

	- namespace cleanup is enabled by default. This can be disabled by
	touch ${sysconfdir}/vservers/.defaults/nonamespace-cleanup

	- all unset ulimits are now reset to infinity, so limits are no longer
	inherited from the host.

	- much improved Gentoo support, thanks to Benedikt Böhm.

version 0.30.211
================

	- novlandev is now the default. To reenable the old behaviour of
	creating VLAN interfaces where used, touch
	${sysconfdir}/vservers/.defaults/interfaces/vlandev.

	- vshelper now works correctly with the kernel legacy support turned
	off.

	- vlogin provides a terminal proxy which allocates a new psuedo-tty
	on the inside when you enter a guest.

	- and lots of other features/fixes, see ChangeLog.


version 0.30.210
================

	- LEGACY BREAKAGE: by default, only recent kernel APIs will be
	supported so you have to make sure to use at least the vs2.0.1
	kernel patch. With older kernels, you have to call 'configure'
	with '--enable-apis=NOLEGACY'. Some programs will not work anymore
	with the default settings; these ones are:

	     + capchroot
	     * chcontext-context
	     * reducecap

	- dlimit support was integrated into the standard vserver
	operations. Thanks to Daniel Hokka Zakrisson for his patch.

	- recent kernel API is now used for setting up the network
	contexts. So it should be now possible to use a kernel compiled
	without CONFIG_VSERVER_LEGACYNET. Thanks to Daniel Hokka Zakrisson
	for his patch.  .


version 0.30.209
================

	- large revamping...
	
	
version 0.23.6
==============

	- fixed '--level' option on 'vserver ... chkconfig'


src/showattr.c
// $Id$

// Copyright (C) 2003 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
// based on showattr.cc by Jacques Gelinas
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2, or (at your option)
// any later version.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "fstool.h"
#include "util.h"

#include <lib/fmt.h>
#include <lib/vserver.h>
#include <lib/vserver-internal.h>

#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <ctype.h>

struct option const
CMDLINE_OPTIONS[] = {
  { "help",     no_argument,  0, CMD_HELP },
  { "version",  no_argument,  0, CMD_VERSION },
#ifdef VC_ENABLE_API_LEGACY
  { "legacy",    no_argument, 0, CMD_LEGACY },
#endif
  { 0,0,0,0 }
};

char const		CMDLINE_OPTIONS_SHORT[] = "Radx";

void
showHelp(int fd, char const *cmd, int res)
{
  WRITE_MSG(fd, "Usage:  ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " [-Radx] [--] <file>*\n\n"
	    " Options:\n"
	    "   -R  ...  recurse through directories\n"
	    "   -a  ...  display files starting with '.' also\n"
	    "   -d  ...  list directories like other files instead of listing\n"
	    "            their content\n"
	    "   -x  ...  do not cross filesystems\n\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");
  exit(res);
}

void
showVersion()
{
  WRITE_MSG(1,
	    "showattr " VERSION " -- shows vserver specific file attributes\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2004 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

void
fixupParams(struct Arguments UNUSED * args, int UNUSED argc)
{
}

static bool
getFlags(char const *name, uint32_t *flags, uint32_t *mask)
{
  xid_t		xid;
  *mask = ~0;
  
  if (vc_get_iattr(name, &xid, flags, mask)==-1) {
    perror("vc_get_iattr()");
    return false;
  }

  return true;
}

bool
handleFile(char const *name, char const *display_name)
{
  bool			res = true;
  char			buf[40];
  char			*ptr = buf;
  uint32_t		flags;
  uint32_t		mask;

  memset(buf, ' ', sizeof buf);

  if (getFlags(name, &flags, &mask)) {
      //                                     1       1       0       0
      //                              fedcba9876543210fedcba9876543210
    static char const	MARKER[33] = ".......x....ciub............mhwa";
    int		i;
    uint32_t 		used_flags = (VC_IATTR_XID|VC_IATTR_ADMIN|
				      VC_IATTR_WATCH|VC_IATTR_HIDE|
				      VC_IATTR_BARRIER|VC_IATTR_IUNLINK|
				      VC_IATTR_IMMUTABLE|VC_IATTR_WRITE|
				      VC_IATTR_COW);

    for (i=0; i<32; ++i) {
      if (used_flags & 1) {
	if (!   (mask  & 1) ) *ptr++ = '-';
	else if (flags & 1)   *ptr++ = toupper(MARKER[31-i]);
	else                  *ptr++ = MARKER[31-i];
      }

      used_flags >>= 1;
      flags      >>= 1;
      mask       >>= 1;
    }
  }      
  else {
    memcpy(buf, "ERR    ", 9);
    res = false;
  }

  Vwrite(1, buf, 10);
  Vwrite(1, display_name, strlen(display_name));
  Vwrite(1, "\n", 1);

  return res;
}
